cloudslang-runtime/src/main/java/io/cloudslang/lang/runtime/steps/AsyncLoopSteps.java

```diff
@@ -0,0 +1,219 @@
+                            @Param(ScoreLangConstants.RUN_ENV) RunEnvironment runEnv,
+                            @Param(EXECUTION_RUNTIME_SERVICES) ExecutionRuntimeServices executionRuntimeServices,
+                            @Param(ScoreLangConstants.NODE_NAME_KEY) String nodeName,
+                            @Param(ExecutionParametersConsts.RUNNING_EXECUTION_PLAN_ID) Long RUNNING_EXECUTION_PLAN_ID,
+                            @Param(ScoreLangConstants.NEXT_STEP_ID_KEY) Long nextStepId,
+                            @Param(ScoreLangConstants.BRANCH_BEGIN_STEP_ID_KEY) Long branchBeginStep,
+                            @Param(ScoreLangConstants.REF_ID) String refId) {
+
+        try {
+            Context flowContext = runEnv.getStack().popContext();
+
+            List<Serializable> splitData = asyncLoopBinding.bindAsyncLoopList(asyncLoopStatement, flowContext, nodeName);
+
+            fireEvent(
+                    executionRuntimeServices,
+                    runEnv,
+                    ScoreLangConstants.EVENT_ASYNC_LOOP_INPUT_END,
+                    "async loop list evaluated",
+                    Pair.of(LanguageEventData.BOUND_ASYNC_LOOP_INPUTS, (Serializable) splitData),
+                    Pair.of(LanguageEventData.levelName.TASK_NAME.name(), nodeName));
+
+            runEnv.putNextStepPosition(nextStepId);
+
+            for (Serializable splitItem : splitData) {
+                RunEnvironment branchRuntimeEnvironment = (RunEnvironment) SerializationUtils.clone(runEnv);
+                Context branchContext = (Context) SerializationUtils.clone(flowContext);
+                branchContext.putVariable(asyncLoopStatement.getVarName(), splitItem);
+
+                updateCallArgumentsAndPushContextToStack(branchRuntimeEnvironment, branchContext, new HashMap<String, Serializable>());
+
+                createBranch(
+                        branchRuntimeEnvironment,
+                        executionRuntimeServices,
+                        RUNNING_EXECUTION_PLAN_ID,
+                        refId,
+                        nextStepId,
+                        branchBeginStep);
+
+                fireEvent(
+                        executionRuntimeServices,
+                        runEnv,
+                        ScoreLangConstants.EVENT_BRANCH_START,
+                        "async loop branch created",
+                        Pair.of(ScoreLangConstants.REF_ID, refId),
+                        Pair.of(ScoreLangConstants.SPLIT_ITEM_KEY, splitItem),
+                        Pair.of(LanguageEventData.levelName.TASK_NAME.name(), nodeName));
+            }
+
+            updateCallArgumentsAndPushContextToStack(runEnv, flowContext, new HashMap<String, Serializable>());
+
+            // forward after the branches are created because begin task method also calls forward
+            runEnv.getExecutionPath().forward();
```

Comment:
we don't forward twice.. when the branches are cloned, they have 0.0 - the branch begin task will forward to 0.1.

at the end of the addBranches method we forward the initial runEnv (will also have 0.1) to remain consistent (they all belong to async loop 0.1)

one problem with this: the events fired before the inital runEnv had 0.0 which should be 0.1 - fixing this in my next push

now all the async loop events have the same path